src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -589,11 +589,31 @@ void MeshForwarder::UpdateRoutes(uint8_t *           aFrame,
 {
     Ip6::Header ip6Header;
     Neighbor *  neighbor;
+    bool        applyOptimization = false;
 
     VerifyOrExit(!aMeshDest.IsBroadcast() && aMeshSource.IsShort());
     SuccessOrExit(GetIp6Header(aFrame, aFrameLength, aMeshSource, aMeshDest, ip6Header));
 
-    Get<AddressResolver>().UpdateCacheEntry(ip6Header.GetSource(), aMeshSource.GetShort());
+    if (!ip6Header.GetSource().IsRoutingLocator() && !ip6Header.GetSource().IsAnycastRoutingLocator() &&
+        Get<NetworkData::Leader>().IsOnMesh(ip6Header.GetSource()) /* only for on mesh address which may require AQ */)
+    {
+        // Thread 1.1 Specification 5.5.2.2:
+        // An FTDs MAY add/update EID-to-RLOC Map Cache entries by inspecting packets being received.
+        if ((Get<Mle::MleRouter>().IsFullThreadDevice() /* FTD */ &&
```

Comment:
How about performing this logic only when `UpdateCacheEntry()` fails?

Reply:
updated. Thanks.